### PR TITLE
ARTEMIS-2787 - Add ability to disable and enable a queue

### PR DESCRIPTION
**Pull request by michaelandrepearce**
Add feature
Add tests
Add docs
Add missing bits noticed in ring-size

### CODE REVIEW

**Comment by clebertsuconic** [comment]
I am trying to understand what disabling a queue would mean. There's already pause on a queue.

**Comment by clebertsuconic** [comment]
The only outcome that I saw from setting enabled or disabled is the boolean changed itself.. being a metadata that serves no purpose to the broker itself.

wouldn't be best to implement a generic meta-data for this kind of purpose?

**Comment by clebertsuconic** [comment]
unless you tell me there's a function for disabling the queue on the broker... I would prefer to serve your feature with a generic metadata, that you set at your own desire. as I couldn't find a function for the broker itself.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/QueueAttributes.java [hunk 1]
```diff
@@ -103,6 +105,8 @@ public void set(String key, String value) {
             setAutoDeleteMessageCount(Long.valueOf(value));
          } else if (key.equals(RING_SIZE)) {
             setRingSize(Long.valueOf(value));
+         } else if (key.equals(ENABLED)) {
+            setRingSize(Long.valueOf(value));
          }
```
Comment (jbertram) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/QueueAttributes.java:110]: Shouldn't this be `setEnabled(Boolean.valueOf(value));`?

**Comment by jbertram** [comment]
@clebertsuconic, look at `QueueImpl.route()`. That's where the functionality is actually implemented.

**Comment by jbertram** [comment]
I think it would be worth making an explicit statement in the documentation that disabling all the queues on an address means that any message sent to that address will be silently dropped. I'm sure this point will be clear to savvy users, but others may not connect the dots as readily.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/QueueAttributes.java (hunk 1 above)
Comment (michaelpearce-gain) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/QueueAttributes.java:110]: yes good catch

**Comment by michaelpearce-gain** [comment]
@clebertsuconic re
`I am trying to understand what disabling a queue would mean. There's already pause on a queue.`

pause only works when no consumers, in this case we need to stop messages entering the queue, even if consumers are attached.

**Comment by michaelpearce-gain** [comment]
@jbertram will add extra line to the docs to mention that.

**Comment by michaelpearce-gain** [comment]
@clebertsuconic its used by the queue to disable it thats what the feature adds, as such its not meta-data.

File: docs/user-manual/en/address-model.md [hunk 2]
```diff
@@ -527,6 +527,34 @@ Open the file `<broker-instance>/etc/broker.xml` for editing.
...
+<addresses>
+   <address name="foo.bar">
+      <multicast>
+         <queue name="orders1" enabled="false"/>
+      </multicast>
+   </address>
+</addresses>
+```
+
+Warning: Disabling all the queues on an address means that any message sent to that address will be silently dropped.
```
Comment (michaelpearce-gain) [inline docs/user-manual/en/address-model.md:555]: @jbertram  added doc in line to be explicit if someone disables all queues.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/QueueAttributes.java (hunk 1 above)
Comment (michaelpearce-gain) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/QueueAttributes.java:110]: fixed

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/addressing/AddressingTest.java [hunk 3]
```diff
@@ -268,6 +268,62 @@ public void testPurgeOnNoConsumersFalse() throws Exception {
       Wait.assertEquals(1, server.locateQueue(queueName)::getMessageCount);
    }
 
+
+   @Test
+   public void testQueueEnabledDisabled() throws Exception {
```
Comment (michaelpearce-gain) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/addressing/AddressingTest.java:273]: @clebertsuconic see here this is the actual functionality test, where enabling / disabling the queue.

(other tests added are just around checking config works and updates etc should reload etc occur)

**Comment by clebertsuconic** [comment]
I missed the QueueImpl::route, I only saw the change the PostOfficeImpl... 

it makes sense as a feature.

**Comment by michaelpearce-gain** [comment]
@clebertsuconic cool :) , anything else needed before merge? I fixed the bit justin spotted, and added the warning to docs regarding if someone disabled all.

**Comment by michaelpearce-gain** [comment]
@clebertsuconic i dont follow sorry. its stored in the queue config, as such if you update the queue config, as like any other queue config change it will be persisted.

**Comment by clebertsuconic** [comment]
@michaelandrepearce queue.pause / queue.resume.. there's an option on the JMX Console method that will store the value on the journal (not on the config)


I was just asking if wouldn't make sense to do the same.

**Comment by michaelpearce-gain** [comment]
so thats taken care of if you update config via jmx console

https://github.com/apache/activemq-artemis/pull/3169/files#diff-1fae8ea746270e6f0b1f0c99a6148ab2

**Comment by michaelpearce-gain** [comment]
i see what you mean, i think holding state somewhere else would be dangerous, we could though do the updae via post office, as such it will update the binding and then queue that way, as such it will go via same route as config change


e.,g. 
         server.getPostOffice().updateQueue(queue.getQueueConfiguration().setEnabled(false));

**Comment by michaelpearce-gain** [comment]
Ill send that in two ticks.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/QueueControlImpl.java [hunk 4]
```diff
@@ -618,6 +618,52 @@ public boolean isPurgeOnNoConsumers() {
...
+   @Override
+   public void enable() throws Exception {
+      if (AuditLogger.isEnabled()) {
+         AuditLogger.enable(queue);
+      }
+      checkStarted();
+
+      clearIO();
+      try {
+         server.getPostOffice().updateQueue(queue.getQueueConfiguration().setEnabled(true));
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/QueueControlImpl.java:645]: now it updates the config, as such persists in the queue binding.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/QueueControlImpl.java:645]: nice one... I will run tests and merge it

**Comment by clebertsuconic** [comment]
@michaelandrepearce / @michaelpearce-gain a test you wrote is failing 100% of the time. Can you take a look?

org.apache.activemq.artemis.tests.integration.client.SessionTest::testCreateQueue

**Comment by michaelpearce-gain** [comment]
@clebertsuconic  fixed , the new case is only valid for new createQueue (not legacy , e.g the new unified createQueue(QueueConfig))

**Comment by clebertsuconic** [comment]
I'm just confirming one non related failure on my last run. if everything ok I will merge this today.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionQueueQueryResponseMessage_V3.java [hunk 5]
```diff
@@ -475,6 +501,11 @@ public boolean equals(Object obj) {
             return false;
       } else if (!ringSize.equals(other.ringSize))
          return false;
+      if (enabled == enabled) {
```
Comment (ehsavoie) [inline artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionQueueQueryResponseMessage_V3.java:504]: This is wrong I think

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionQueueQueryResponseMessage_V3.java:504]: good spot, https://github.com/apache/activemq-artemis/pull/3174 fix here.

**Pull request closed without merging**